src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/Server/SmiEventSink_Default.cs

```diff
@@ -266,21 +266,12 @@ internal override void DefaultDatabaseChanged(string databaseName)
         // Called for messages and errors (ERROR and INFO tokens)
         internal override void MessagePosted(int number, byte state, byte errorClass, string server, string message, string procedure, int lineNumber)
         {
+            var serverValue = (null != server) ? server : "<null>";
+            var messageBody = (null != message) ? message : "<null>";
+            var proc = (null != procedure) ? procedure : "<null>";
             if (null == _parent)
             {
-                if (Bid.AdvancedOn)
-                {
-                    Bid.Trace("<sc.SmiEventSink_Default.MessagePosted|ADV> %d#, number=%d state=%d errorClass=%d server='%ls' message='%ls' procedure='%ls' linenumber=%d.\n",
-                                0,
-                                number,
-                                state,
-                                errorClass,
-                                (null != server) ? server : "<null>",
-                                (null != message) ? message : "<null>",
-                                (null != procedure) ? procedure : "<null>",
-                                lineNumber
-                                );
-                }
+                SqlClientEventSource._log.Trace($"<sc.SmiEventSink_Default.MessagePosted|ADV> {0}#, number={number} state={state} errorClass={errorClass} server='{serverValue}' message='{messageBody}' procedure='{proc}' linenumber={lineNumber}.\n");
```

Comment:
From a performance perspective this pattern is going to suck. The call has to format the value to pass it as a parameter to the method which will always cause allocation of the string which can't be reused to recycled. Then the method is going to check if tracing is enabled and (most of the time) return early without using the string. Even thoough it's much more verbose and understandably annoying to have to copy paste everywhere it's much better for consumers to do the check for tracing being enabled before formatting up the string for output.

Reply:
Thanks @Wraith2  for catching this. I will replace the local variables with null-coalescing operator which will be inside the EventSource line. the checking is done inside the SqlClientEventSource class. For "Trace" it checks for Event level and Keywords which is Informational and Trace(0x0004) accordingly.

Reply:
The same pattern applies everywhere that output needs to be constructed unfortunately. I've had the same experience with `DiagnosticSource` which uses the same enablement check pattern. As you say there are ways to try and work around it using nullable structs and other language trickery but without being able to use later language versions the basic pattern pretty much has to be:

```csharp
if (TraceEventSource.IsTraceEnabled)
{
    TraceEventSource.Trace($"format message here");
}
```
or the allocations always end up happening even in the disabled case.

Reply:
So, if I change it to  `_log.Trace($"<sc.SmiEventSink_Default.MessagePosted|ADV> {0}#, number={number} state={state} errorClass={errorClass} server='{server ?? "<null>"}' message='{message ?? "<null>"}' procedure='{procedure ?? "<null>"}' linenumber={lineNumber}.\n");` it wont work? The IsEnabled is applied for Trace  ```
if (_log.IsEnabled(EventLevel.Informational, Keywords.Trace))
            {
                WriteEvent(TraceEventId, message);
            }
```

Reply:
The driver was able to accept this pattern.

Reply:
The code is correct c#. it'll work. It will always allocate the string to pass it as a parameter to the Trace function even though in the majority of cases the IfEnabled call will return false and the string will never be used. It's poor from a performance/GC perspective not a correctness perspective.

Reply:
My suggestion would be:

```csharp
internal class SqlClientEventSource : EventSource
{
	// most of the class as already written

	internal bool IsTraceEnabled => _log.IsEnabled(EventLevel.Informational, Keywords.Trace);

        [Event(TraceEventId, Level = EventLevel.Informational, Keywords = Keywords.Trace)] 
        internal void Trace(string message) 
        { 
            WriteEvent(TraceEventId, message); 
        }
}
```
then to use it
```csharp
// first check if we need to call trace
if (SqlClientEventSource.IsTraceEnabled)
{
	// only if the trace flag is enabled, allocate the string and pass it to the method
	SqlClientEventSource.Trace($"<sc.SmiEventSink_Default.MessagePosted|ADV> {0}#, number={number} state={state} errorClass={errorClass} server='{server ?? "<null>"}' message='{message ?? "<null>"}' procedure='{procedure ?? "<null>"}' linenumber={lineNumber}.\n");
}
```